lib/elixir/test/elixir/module/types/types_test.exs

```diff
@@ -52,4 +105,388 @@ defmodule Module.Types.TypesTest do
+                 is_binary(y)
+
+             where "x" was given the type integer() in:
+
+                 # types_test.ex:1
+                 is_integer(x)
+             """
+    end
+
+    test "only show relevant traces in warning" do
+      string = warning([x = y, z], [is_integer(x) and is_binary(y) and is_boolean(z)], {x, y, z})
+
+      assert string == """
+             incompatible types:
+
+                 integer() !~ binary()
+
+             in expression:
+
+                 # types_test.ex:1
+                 is_integer(x) and is_binary(y) and is_boolean(z)
+
+             where "y" was given the same type as "x" in:
+
+                 # types_test.ex:1
+                 x = y
+
+             where "y" was given the type binary() in:
+
+                 # types_test.ex:1
+                 is_binary(y)
+
+             where "x" was given the type integer() in:
+
+                 # types_test.ex:1
+                 is_integer(x)
+             """
+    end
+
+    test "check body" do
+      string = warning([x], [is_integer(x)], :foo = x)
+
+      assert string == """
+             incompatible types:
+
+                 integer() !~ :foo
+
+             in expression:
+
+                 # types_test.ex:1
+                 :foo = x
+
+             where "x" was given the type integer() in:
+
+                 # types_test.ex:1
+                 is_integer(x)
+
+             where "x" was given the type :foo in:
+
+                 # types_test.ex:1
+                 :foo = x
+             """
+    end
+
+    test "check binary" do
+      string = warning([foo], [is_binary(foo)], <<foo>>)
+
+      assert string == """
+             incompatible types:
+
+                 binary() !~ integer()
+
+             in expression:
+
+                 # types_test.ex:1
+                 <<foo>>
+
+             where "foo" was given the type binary() in:
+
+                 # types_test.ex:1
+                 is_binary(foo)
+
+             where "foo" was given the type integer() in:
+
+                 # types_test.ex:1
+                 <<foo>>
+
+             HINT: all expressions given to binaries are assumed to be of type \
+             integer() unless said otherwise. For example, <<expr>> assumes "expr" \
+             is an integer. Pass a modifier, such as <<expr::float>> or <<expr::binary>>, \
+             to change the default behaviour.
+             """
+
+      string = warning([foo], [is_binary(foo)], <<foo::integer>>)
+
+      assert string == """
+             incompatible types:
+
+                 binary() !~ integer()
+
+             in expression:
+
+                 # types_test.ex:1
+                 <<foo::integer()>>
+
+             where "foo" was given the type binary() in:
+
+                 # types_test.ex:1
+                 is_binary(foo)
+
+             where "foo" was given the type integer() in:
+
+                 # types_test.ex:1
+                 <<foo::integer()>>
```

Comment:
How come we print as `<<foo::integer()>>` instead of `<<foo::integer>>`?

Reply:
We need to fix `Macro.to_string` to skip parens on the right side of `::` inside `<<>>` for the built-in identifiers.